packages/core/src/types.ts

```diff
@@ -20,9 +20,12 @@ export type Compute<A extends any> = { [K in keyof A]: A[K] } & unknown;
 export type Prop<T, K> = K extends keyof T ? T[K] : never;
 export type Values<T> = T[keyof T];
 export type Merge<M, N> = Omit<M, keyof N> & N;
+// TODO: replace in v5 with:
+// export type IndexByType<T extends { type: string }> = { [E in T as E['type']]: E; };
 export type IndexByType<T extends { type: string }> = {
-  [K in T['type']]: Extract<T, { type: K }>;
+  [K in T['type']]: T extends any ? (K extends T['type'] ? T : never) : never;
```

Comment:
I wonder why `T extends any` makes this work 🧐

Reply:
https://github.com/statelyai/eng-blog/pull/81/files :P

This is just a way to make this conditional type distributive. Conditional types are distributive when they contain a "naked" type param:
```ts
type IsStringDistributive<T> = T extends string ? 1 : 0

// this will run the conditional type over each member of this union type
type Distributes = IsStringDistributive<string | number>
//   ^? 1 | 0

type IsStringNonDistributive<T> = [T] extends [string] ? 1 : 0

// this will run the conditional type on the union type "as is"
type DoesntDistribute = IsStringNonDistributive<string | number>
//   ^? 0
```

[TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAkgzgZWAJwJYDsDmARVcWoBGArsKgG4QA8AKgHxQC8UNUEAHsBOgCZxT40WKAH4oARigAuKAAYAUPNCQouQUVIR+zeEiE48BEmUpV1wgD5R0xALaEIyOvID0LqB4B6Ixcui6CLAA5AHt0NSNSCmp6JigAbRoAXTZObj4E80wUsUkZBT9VEK10YAi0Y2gdREDMUPDDCqjTLKgrG3tHZzcPKG8gA)
[TS playground with Omit variants](https://www.typescriptlang.org/play?&fbclid=IwAR3oPnf0KrLaCVZdkcZcnC4rxtz2WpV9SpsGa4XUNUHny5UwXE8BYISJDMw#code/C4TwDgpgBAsiDyBbAlsAPAFQDRQNID4oBeKAbygG0AFKZAOygFEAPAYwBsBXAEwjQGsIIAPYAzKNjz4AugC4J1aQG4oAXwBQ60JCgARZAGdgAJ2QAjTsGQA3CElSYcBYhKgRmwCHW4GoAQzoQKAB+WAQUdElneToIW2NNbWg4AFU6ZGEGEnIk+SNTOgBzFQCDAHcIYxjORDNKnGFgAAtK+TNhYXYIALUoAB8yKFyofPpi-zpy1qg6GrrjFVZhRDAu5nkAJQgl4240UaKcTjp+OmEyukINLXBoADlM-VGLK1stg052YBc4e3RU9KZHAAciSwPw6gA9JCoFAAHrBRK3PSGEzmSw2CDvT7fEhPNEvTF-NAAjJ0EFgiHQ2EIoA)

So this particular type here just checks if the current `K` extends any of the `T['type']` and it will match even if `T['type']` is a union. Whereas the previous version would fail because
```ts
'FOO' | 'BAR' extends 'FOO' ? 1 : 0 // 0
```